src/client/lib/authenticated-page.tsx

```diff
@@ -0,0 +1,27 @@
+import { useEffect } from "react";
+import { useRouter } from "next/router";
+import { useSession } from "next-auth/react";
+
+export const requiresLogin =
+  <Props extends object>(WrappedPage: React.ComponentType<Props>) =>
+  // eslint-disable-next-line react/display-name
+  (props: Props) => {
+    const router = useRouter();
+    const { status } = useSession();
+
+    // Include any public pathnames here
+    const publicPaths = ["/signInPage"];
+    const pathIsProtected = !publicPaths.includes(router.pathname);
+
+    useEffect(() => {
+      if (status === "unauthenticated" && pathIsProtected) {
+        void router.push("/");
+      }
+    });
```

Comment:
Let's try it! TS prompted me to add all three dependencies in the useEffect, so it will trigger on changes in router, status, and pathIsProtected after my changes, but that's still better than every render.